larq/testing_utils.py

```diff
@@ -27,3 +32,172 @@ def get_small_bnn_model(input_dim, num_hidden, output_dim):
+            if e is None:
+                input_data_shape[i] = np.random.randint(1, 4)
+        input_data = 10 * np.random.random(input_data_shape)
+        if input_dtype[:5] == "float":
+            input_data -= 0.5
+        input_data = input_data.astype(input_dtype)
+    elif input_shape is None:
+        input_shape = input_data.shape
+    if input_dtype is None:
+        input_dtype = input_data.dtype
+    if expected_output_dtype is None:
+        expected_output_dtype = input_dtype
+
+    # instantiation
+    kwargs = kwargs or {}
+    layer = layer_cls(**kwargs)
+
+    # test get_weights , set_weights at layer level
+    weights = layer.get_weights()
+    layer.set_weights(weights)
+
+    # test and instantiation from weights
+    if "weights" in inspect.getargspec(layer_cls.__init__):
+        kwargs["weights"] = weights
+        layer = layer_cls(**kwargs)
+
+    # test in functional API
+    x = tf.keras.layers.Input(shape=input_shape[1:], dtype=input_dtype)
+    y = layer(x)
+    if tf.keras.backend.dtype(y) != expected_output_dtype:
+        raise AssertionError(
+            "When testing layer %s, for input %s, found output "
+            "dtype=%s but expected to find %s.\nFull kwargs: %s"
+            % (
+                layer_cls.__name__,
+                x,
+                tf.keras.backend.dtype(y),
+                expected_output_dtype,
+                kwargs,
+            )
+        )
+    # check shape inference
+    model = tf.keras.models.Model(x, y)
+    expected_output_shape = tuple(
+        layer.compute_output_shape(tf.TensorShape(input_shape)).as_list()
+    )
+    actual_output = model.predict(input_data)
+    actual_output_shape = actual_output.shape
+    for expected_dim, actual_dim in zip(expected_output_shape, actual_output_shape):
+        if expected_dim is not None:
+            if expected_dim != actual_dim:
+                raise AssertionError(
+                    "When testing layer %s, for input %s, found output_shape="
+                    "%s but expected to find %s.\nFull kwargs: %s"
+                    % (
+                        layer_cls.__name__,
+                        x,
+                        actual_output_shape,
+                        expected_output_shape,
+                        kwargs,
+                    )
+                )
+    if expected_output is not None:
+        np.testing.assert_allclose(actual_output, expected_output, rtol=1e-3)
+
+    # test serialization, weight setting at model level
+    model_config = model.get_config()
+    recovered_model = tf.keras.models.Model.from_config(model_config)
+    if model.weights:
+        weights = model.get_weights()
+        recovered_model.set_weights(weights)
+        output = recovered_model.predict(input_data)
+        np.testing.assert_allclose(output, actual_output, rtol=2e-3)
+
+    # Recreate layer to prevent layer metrics from being configured multiple times.
+    layer = layer_cls(**kwargs)
```

Comment:
This is the line, why we need to fork this function. In general I think it is not too bad to maintain our own layer tests to have full control.